
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Maddy Duran">
      
      
        <link rel="canonical" href="https://cole-trapnell-lab.github.io/platt/deg/">
      
      
        <link rel="prev" href="../graphs/">
      
      
        <link rel="next" href="../dveg/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.11">
    
    
      
        <title>DEGs - Platt</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.4af4bdda.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="trapnell" data-md-color-primary="indigo" data-md-color-accent="teal">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#differential-expression-analysis-in-platt" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Platt" class="md-header__button md-logo" aria-label="Platt" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Platt
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              DEGs
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/cole-trapnell-lab/platt" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Platt" class="md-nav__button md-logo" aria-label="Platt" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Platt
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/cole-trapnell-lab/platt" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Background
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../install/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../kinetics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Kinetics
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../graphs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Graphs
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    DEGs
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    DEGs
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#running-degs-over-a-graph" class="md-nav__link">
    <span class="md-ellipsis">
      Running DEGs over a graph
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-degs-within-each-perturbation" class="md-nav__link">
    <span class="md-ellipsis">
      Running DEGs within each perturbation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      References
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../dveg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    DvEGs
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Plotting
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#running-degs-over-a-graph" class="md-nav__link">
    <span class="md-ellipsis">
      Running DEGs over a graph
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-degs-within-each-perturbation" class="md-nav__link">
    <span class="md-ellipsis">
      Running DEGs within each perturbation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      References
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="differential-expression-analysis-in-platt">Differential expression analysis in Platt</h1>
<h2 id="running-degs-over-a-graph">Running DEGs over a graph</h2>
<p>Finding regulatory genes with fate-restricted patterns may help identify new genetic requirements of cell types. For example, “terminal selector” and “multilineage priming (MLP)” genes might activate specific fates, genes expressed in progenitors only might be required for maintaining the progenitor states, and genes excluded from certain fates might repress that fate <a href="https://cole-trapnell-lab.github.io/platt/deg/#references">1-4</a>. We can systematically identify these patterns by computing differential expression across our graphs and then classify these patterns based on a set of defined rules.  </p>
<p><img alt="" src="../assets/degs_over_graph.png" /></p>
<p><em>See an explanation of gene patterns <a href="https://cole-trapnell-lab.github.io/platt/patterns/">here</a>:</em></p>
<p>The function <code>compare_genes_within_state_graph()</code>:</p>
<ul>
<li><code>ccs</code>- a Hooke <code>cell_count_set</code> object</li>
<li><code>graph</code> - an <code>igraph</code> object</li>
<li><code>gene_ids</code> - a list of genes to subset the analysis to </li>
<li><code>cores</code> - number of cores </li>
</ul>
<pre><code>pf_graph_degs = compare_genes_over_graph(pf_ccs,
                                         pf_cell_state_graph@graph, 
                                         cores = 4)
</code></pre>
<p>The output of this table will look like this: </p>
<table>
<thead>
<tr>
<th>cell_state</th>
<th>gene_class_scores</th>
</tr>
</thead>
<tbody>
<tr>
<td>pectoral fin condensate</td>
<td><code>&lt;tibble [14,899 × 5]&gt;</code></td>
</tr>
<tr>
<td>pectoral fin distal mesenchyme</td>
<td><code>&lt;tibble [14,899 × 5]&gt;</code></td>
</tr>
<tr>
<td>pectoral fin central cells</td>
<td><code>&lt;tibble [14,899 × 5]&gt;</code></td>
</tr>
<tr>
<td>pectoral fin bud mesoderm</td>
<td><code>&lt;tibble [14,899 × 5]&gt;</code></td>
</tr>
<tr>
<td>pectoral fin cleithrum</td>
<td><code>&lt;tibble [14,899 × 5]&gt;</code></td>
</tr>
<tr>
<td>pectoral fin bud progenitor</td>
<td><code>&lt;tibble [14,899 × 5]&gt;</code></td>
</tr>
</tbody>
</table>
<p>To unnest the dataframe: </p>
<pre><code>pf_graph_degs %&gt;% 
  tidyr::unnest(gene_class_scores) %&gt;% 
  filter(pattern_activity_score &gt; 1) %&gt;%
  filter(interpretation == &quot;Selectively activated&quot;)

</code></pre>
<table>
<thead>
<tr>
<th>cell_state</th>
<th>gene_id</th>
<th>data</th>
<th>interpretation</th>
<th>pattern_activity_score</th>
<th>gene_short_name</th>
</tr>
</thead>
<tbody>
<tr>
<td>pectoral fin condensate</td>
<td>ENSDARG000000062…</td>
<td><code>&lt;tibble&gt;</code></td>
<td>Selectively activated</td>
<td>1.02</td>
<td>ell2</td>
</tr>
<tr>
<td>pectoral fin condensate</td>
<td>ENSDARG000000099…</td>
<td><code>&lt;tibble&gt;</code></td>
<td>Selectively activated</td>
<td>2.16</td>
<td>slc38a5a</td>
</tr>
<tr>
<td>pectoral fin condensate</td>
<td>ENSDARG000000106…</td>
<td><code>&lt;tibble&gt;</code></td>
<td>Selectively activated</td>
<td>1.86</td>
<td>clic2</td>
</tr>
<tr>
<td>pectoral fin condensate</td>
<td>ENSDARG000000116…</td>
<td><code>&lt;tibble&gt;</code></td>
<td>Selectively activated</td>
<td>1.19</td>
<td>slc26a2</td>
</tr>
<tr>
<td>pectoral fin condensate</td>
<td>ENSDARG000000124…</td>
<td><code>&lt;tibble&gt;</code></td>
<td>Selectively activated</td>
<td>3.77</td>
<td>col11a2</td>
</tr>
<tr>
<td>pectoral fin condensate</td>
<td>ENSDARG000000309…</td>
<td><code>&lt;tibble&gt;</code></td>
<td>Selectively activated</td>
<td>2.00</td>
<td>mybl1</td>
</tr>
</tbody>
</table>
<p>We can check some of these markers by plotting them either in the UMAP space:</p>
<pre><code>plot_cells(pf_ccs@cds, genes = c(&quot;emilin3a&quot;))
</code></pre>
<p><img alt="" src="../assets/pf_emilin3a.png" width="50%" /></p>
<p>... or on our platt graph:</p>
<pre><code>plot_gene_expr(pf_cell_state_graph, genes = c(&quot;emilin3a&quot;))
</code></pre>
<p><img alt="" src="../assets/pf_graph_emilin3a.png" width="75%" /></p>
<p><em>For more information about plotting on a Platt graph, see our <a href="https://cole-trapnell-lab.github.io/platt/plotting">plotting page</a>.</em></p>
<h2 id="running-degs-within-each-perturbation">Running DEGs within each perturbation</h2>
<p>The function <code>compare_genes_within_state_graph()</code>: </p>
<ul>
<li><code>ccs</code>- a Hooke <code>cell_count_set</code> object</li>
<li><code>perturbation_col</code> - column name of the perturbations</li>
<li><code>control_ids</code> - list of control ids </li>
<li><code>cell_groups</code> - subset of cell groups to run DEGs on </li>
<li><code>perturbations</code> - defaults to perturbation</li>
<li><code>cores</code></li>
</ul>
<p>For this example we will be using a subset of the skeletal muscle, for which we have 510,093 reference cells: </p>
<pre><code>platt:::plot_annotations(muscle_state_graph, plot_labels = T, node_size = 4)

</code></pre>
<p><img alt="" src="../assets/muscle_anno.png" width="75%" /></p>
<pre><code>genes_within_cell_state = compare_genes_within_state_graph(ccs, 
                                                           perturbation_col = &quot;gene_target&quot;, 
                                                           control_ids = c(&quot;ctrl-inj&quot;), 
                                                           perturbations = c(&quot;tbx16&quot;, &quot;tbx16-msgn1&quot;, &quot;tbx16-tbx16l&quot;),
                                                           cores = 6)

genes_within_cell_state %&gt;% head()

</code></pre>
<table>
<thead>
<tr>
<th>cell_group</th>
<th>genes_within_cell_group</th>
</tr>
</thead>
<tbody>
<tr>
<td>paraxial mesoderm (tbx16+)</td>
<td><code>&lt;tibble [33,189 × 15]&gt;</code></td>
</tr>
<tr>
<td>paraxial mesoderm (pax3+, pax7-)</td>
<td><code>&lt;tibble [33,189 × 15]&gt;</code></td>
</tr>
<tr>
<td>fast-committed myocyte, pre-fusion</td>
<td><code>&lt;tibble [33,189 × 15]&gt;</code></td>
</tr>
<tr>
<td>fast-committed myocyte, fusing (pcdh7b+)</td>
<td><code>&lt;tibble [33,189 × 15]&gt;</code></td>
</tr>
<tr>
<td>head and neck mesoderm (pax3+, pax7+)</td>
<td><code>&lt;tibble [33,189 × 15]&gt;</code></td>
</tr>
<tr>
<td>head and neck mesoderm, migrating</td>
<td><code>&lt;tibble [33,189 × 15]&gt;</code></td>
</tr>
</tbody>
</table>
<p>The results are nested by cell type. To look at a specific knockout term, you can unnest the dataframe and filter: </p>
<pre><code>genes_within_cell_state = genes_within_cell_state %&gt;% tidyr::unnest(genes_within_cell_group)
genes_within_cell_state %&gt;% filter(term == &quot;tbx16,msgn1&quot;) %&gt;% tidyr::unnest(perturb_effects)

</code></pre>
<table>
<thead>
<tr>
<th>term</th>
<th>mean_log_sf</th>
<th>ctrl_log_sf</th>
<th>detected_genes</th>
<th>ctrl_detected_genes</th>
<th>id</th>
<th>perturb_to_ctrl_raw_lfc</th>
<th>perturb_to_ctrl_raw_lfc_se</th>
<th>perturb_to_ctrl_raw_p_value</th>
<th>perturb_to_ctrl_shrunken_lfc</th>
<th>perturb_to_ctrl_shrunken_lfc_se</th>
<th>perturb_to_ctrl_p_value</th>
<th>effect_skew</th>
<th>log_mean_expression</th>
<th>coefficient_mode</th>
<th>cell_group</th>
</tr>
</thead>
<tbody>
<tr>
<td>tbx16,msgn1</td>
<td>3.955547</td>
<td>2.302981</td>
<td>6042.273</td>
<td>2235.279</td>
<td>ENSDARG00000000002</td>
<td>-0.08138754</td>
<td>0.8304913</td>
<td>0.46096644</td>
<td>-0.007389409</td>
<td>0.25044809</td>
<td>0.7034408</td>
<td>-2.799458</td>
<td>-5.560817</td>
<td>-0.02203216</td>
<td>paraxial mesoderm (tbx16+)</td>
</tr>
<tr>
<td>tbx16,msgn1</td>
<td>3.955547</td>
<td>2.302981</td>
<td>6042.273</td>
<td>2235.279</td>
<td>ENSDARG00000000018</td>
<td>0.88256012</td>
<td>0.3981750</td>
<td>0.01332819</td>
<td>0.314546764</td>
<td>0.34338591</td>
<td>0.3002582</td>
<td>-2.799458</td>
<td>-3.133384</td>
<td>-0.02203216</td>
<td>paraxial mesoderm (tbx16+)</td>
</tr>
<tr>
<td>tbx16,msgn1</td>
<td>3.955547</td>
<td>2.302981</td>
<td>6042.273</td>
<td>2235.279</td>
<td>ENSDARG00000000019</td>
<td>0.04861553</td>
<td>0.2461786</td>
<td>0.42172566</td>
<td>0.014027290</td>
<td>0.13301079</td>
<td>0.7339044</td>
<td>-2.799458</td>
<td>-2.514816</td>
<td>-0.02203216</td>
<td>paraxial mesoderm (tbx16+)</td>
</tr>
<tr>
<td>tbx16,msgn1</td>
<td>3.955547</td>
<td>2.302981</td>
<td>6042.273</td>
<td>2235.279</td>
<td>ENSDARG00000000068</td>
<td>0.03918378</td>
<td>0.3448263</td>
<td>0.45476421</td>
<td>0.008933895</td>
<td>0.16490703</td>
<td>0.7267108</td>
<td>-2.799458</td>
<td>-3.313506</td>
<td>-0.02203216</td>
<td>paraxial mesoderm (tbx16+)</td>
</tr>
<tr>
<td>tbx16,msgn1</td>
<td>3.955547</td>
<td>2.302981</td>
<td>6042.273</td>
<td>2235.279</td>
<td>ENSDARG00000000069</td>
<td>0.22385107</td>
<td>0.2309466</td>
<td>0.16620371</td>
<td>0.076271650</td>
<td>0.15154521</td>
<td>0.5804131</td>
<td>-2.799458</td>
<td>-2.281009</td>
<td>-0.02203216</td>
<td>paraxial mesoderm (tbx16+)</td>
</tr>
<tr>
<td>tbx16,msgn1</td>
<td>3.955547</td>
<td>2.302981</td>
<td>6042.273</td>
<td>2235.279</td>
<td>ENSDARG00000000086</td>
<td>-0.04488597</td>
<td>0.1612511</td>
<td>0.39036776</td>
<td>-0.014688242</td>
<td>0.09378365</td>
<td>0.7574802</td>
<td>-2.799458</td>
<td>-1.528009</td>
<td>-0.02203216</td>
<td>paraxial mesoderm (tbx16+)</td>
</tr>
</tbody>
</table>
<p><em>Pax3a</em> is a MLP gene observed in paraxial mesoderm progenitors as they commit to either head and neck mesoderm or fast muscle fates. </p>
<pre><code>plot_gene_expr(muscle_state_graph, genes = c(&quot;pax3a&quot;), node_size = 4, plot_labels = F) + 
  theme(legend.position = &quot;right&quot;)
</code></pre>
<p><img alt="" src="../assets/muscle_pax3a.png" width="75%" /></p>
<p>In the tbx16-msgn1 crispants, all of which fail to generate skeletal muscle properly, <em>pax3a</em>, 
a key regulator of myogenesis was differentially expressed in myogenic progenitors. </p>
<table>
<thead>
<tr>
<th>term</th>
<th>cell_group</th>
<th>gene_short_name</th>
<th>perturb_to_ctrl_shrunken_lfc</th>
<th>perturb_to_ctrl_p_value</th>
</tr>
</thead>
<tbody>
<tr>
<td>tbx16,msgn1</td>
<td>paraxial mesoderm (tbx16+)</td>
<td>pax3a</td>
<td>-1.0584180</td>
<td>1.397403e-08</td>
</tr>
<tr>
<td>tbx16,msgn1</td>
<td>paraxial mesoderm (pax3+, pax7-)</td>
<td>pax3a</td>
<td>-0.7609017</td>
<td>2.913705e-05</td>
</tr>
</tbody>
</table>
<p>... we can plot the DEG fold change of <em>pax3a</em> in the tbx16-msgn1 mutant on the platt graph: </p>
<pre><code>plot_degs(muscle_state_graph, tbx16_degs %&gt;% 
            filter(term == &quot;tbx16,msgn1&quot;, gene_short_name == &quot;pax3a&quot;), node_size = 4.5)
</code></pre>
<p><img alt="" src="../assets/muscle_pax3a_deg.png" width="75%" /></p>
<p><em>For more information about plotting on a Platt graph, see our <a href="https://cole-trapnell-lab.github.io/platt/plotting">plotting page</a>.</em></p>
<h2 id="references">References</h2>
<ol>
<li>Hobert, O. Terminal selectors of neuronal identity. Curr. Top. Dev. Biol. 116, 455–475 (2016).</li>
<li>Laslo, P. et al. Multilineage transcriptional priming and determination of alternate hematopoietic cell fates. Cell 126, 755–766 (2006).</li>
<li>Qiu, C. et al. Systematic reconstruction of cellular trajectories across mouse embryogenesis. Nat. Genet. 54, 328–341 (2022).</li>
<li>Packer, J. S. et al. A lineage-resolved molecular atlas of C. elegans embryogenesis at single-cell resolution. Science 365, (2019).</li>
</ol>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy 2025 <a href="https://github.com/cole-trapnell-lab"  target="_blank" rel="noopener">Cole Trapnell</a>

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.code.copy"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>